### PR TITLE
fix(drive)!: versioning issues on epoch change

### PR DESCRIPTION
**Pull request by shumkov**
## Issue being fixed or feature implemented

There are multiple versioning issues fixed in this PR:
- We mostly use the version of the previous epoch to process the first block of a new epoch. A new version is still used for a few calls. This can lead to multiple problems, especially if we change fee logic in the new version. We should use only one version for all blocks within the epoch.
- A version proposal is missing for the first block of epoch
- The protocol version in a block header is not updating which can lead to confusion and problems in the future when we need to know what version we should use (or we used) to process the block.

## What was done?

- Updated Platform to Tenderdash 0.14.0-dev.5 that supports the setting of the block protocol version
- Set protocol version to block header on the "prepare proposal" and validate this version on the "process proposal"
- Fixed using a mix of previous and next versions to process the first block of an epoch. Now it uses only the new protocol version. Note: epoch change to determine the version to run block proposal still uses the previous block protocol version.
- Fixed missing version proposal vote for the first block of an epoch.
- Added verbose panic message in case if new protocol version is not supported. Such as a warning if the next epoch protocol version is not supported.
- Fixed `gather_epoch_info` returns v0 instead of versioned `EpochInfo`
- Removed storing of the next protocol version since it will be removed in #1778
- Versioned `protocol_version_upgrade_percentage_needed` constant
- Allowed to modify platform test version mocks for testing
- Removed side effects on calling `versions_passing_threshold`

## How Has This Been Tested?

Updated tests to reflect changes, including an additional version proposal per epoch. 

## Breaking Changes

Previously created state won't be compatible:
- Now we count and persist a version proposal vote on the epoch change.
- Protocol version for the first block of an epoch might be different.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/rs-drive/src/drive/cache/protocol_version.rs [hunk 1]
```diff
@@ -56,33 +55,30 @@ impl ProtocolVersionsCache {
...
-        &mut self,
-        required_upgraded_hpns: u64,
-    ) -> Vec<ProtocolVersion> {
-        self.needs_wipe = true;
+    /// Collect versions passing threshold
+    pub fn versions_passing_threshold(&self, required_upgraded_hpmns: u64) -> Vec<ProtocolVersion> {
         let mut cache = self.global_cache.clone();
 
-        cache.extend(self.block_cache.drain());
+        cache.extend(self.block_cache.iter());
```
Comment (QuantumExplorer) [inline packages/rs-drive/src/drive/cache/protocol_version.rs:77]: I think drain is better so we don't need to clone elements.

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/abci/handler/finalize_block.rs [hunk 2]
```diff
@@ -32,7 +35,11 @@ where
             "block execution context must be set in block begin handler for finalize block",
         )))?;
 
-    let platform_version = app.platform().state.load().current_platform_version()?;
+    // Use current block protocol version
+    let protocol_version = block_execution_context
+        .block_state_info()
+        .protocol_version();
+    let platform_version = PlatformVersion::get(protocol_version)?;
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/abci/handler/finalize_block.rs:42]: You right! It will be the same as version from block platform state

File: packages/rs-drive/src/drive/cache/protocol_version.rs [hunk 3]
```diff
@@ -56,33 +55,28 @@ impl ProtocolVersionsCache {
...
     /// Merge block cache to global cache
     pub fn merge_block_cache(&mut self) {
-        if self.needs_wipe {
-            self.global_cache.clear();
-            self.block_cache.clear();
-            self.needs_wipe = false;
-        } else {
-            self.global_cache.extend(self.block_cache.drain());
-        }
+        self.global_cache.extend(self.block_cache.drain());
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive/src/drive/cache/protocol_version.rs:58]: You right. Fixed.

File: packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/check_for_desired_protocol_upgrade/v0/mod.rs [hunk 4]
```diff
@@ -13,12 +13,12 @@ const PROTOCOL_VERSION_UPGRADE_PERCENTAGE_NEEDED: u64 = 75;
 impl<C> Platform<C> {
     /// checks for a network upgrade and resets activation window
     /// this should only be called on epoch change
-    /// this will change backing state, but does not change drive cache
-    pub(super) fn check_for_desired_protocol_upgrade_v0(
+    /// this will change backing state and drive cache
+    pub(super) fn check_for_desired_protocol_upgrade_and_reset_v0(
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/check_for_desired_protocol_upgrade/v0/mod.rs]: Yes, I totally agree, but it did before. I split read and write logic into different methods to resolve this confusion

File: packages/rs-drive/src/drive/cache/protocol_version.rs (hunk 1 above)
Comment (shumkov) [reply packages/rs-drive/src/drive/cache/protocol_version.rs:77]: In this case this method will alter cache. Getter shouldn't alter cache. How many elements you expect in block cache? I think with such small number there is no difference and it happen only on epoch change

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs [hunk 5]
```diff
@@ -45,16 +48,31 @@ where
...
+            last_committed_platform_version,
         )?;
 
-        match platform_version
+        // Determine a protocol version for this block
+        let block_platform_version = if epoch_info.is_epoch_change_but_not_genesis() {
+            // Switch to next proposed platform version if we are on the first block of the new epoch
+            // This version must be set to the state as current one during block processing
+            let protocol_version = platform_state.next_epoch_protocol_version();
+            PlatformVersion::get(protocol_version)?
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs:69]: This is where things will break if node didn't update.

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs [hunk 6]
```diff
@@ -63,10 +81,10 @@ where
             0 => self.run_block_proposal_v0(
                 block_proposal,
                 known_from_us,
-                epoch_info.into(),
+                epoch_info,
```
Comment (shumkov) [inline packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs:84]: We shouldn't pass EpochInfo from previous version to new version of run block proposal

File: packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/upgrade_protocol_version/v0/mod.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            self.check_for_desired_protocol_upgrade(hpmn_list_len, platform_version)?;
+
+        if let Some(protocol_version) = next_epoch_protocol_version {
+            block_platform_state.set_next_epoch_protocol_version(protocol_version);
+        }
+
+        // Since we are starting a new epoch we need to drop previously
+        // proposed versions
+
+        let mut protocol_versions_counter = self.drive.cache.protocol_versions_counter.write();
```
Comment (shumkov) [inline packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/upgrade_protocol_version/v0/mod.rs:43]: What will happen if we have multiple rounds? We are also querying from the global cache.

Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/upgrade_protocol_version/v0/mod.rs]: Fixed

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs (hunk 5 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs]: Added panic

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs [hunk 8]
```diff
@@ -229,54 +257,15 @@ where
...
-        }
+        // Update block platform state with current and next epoch protocol versions
+        // if it was proposed
+        self.upgrade_protocol_version(
+            &epoch_info,
+            last_committed_platform_state,
+            &mut block_platform_state,
+            transaction,
+            platform_version,
+        )?;
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs]: Done

File: packages/rs-drive-abci/src/config.rs [hunk 9]
```diff
@@ -406,6 +395,9 @@ pub struct PlatformTestConfig {
     pub block_signing: bool,
     /// Block signature verification
     pub block_commit_signature_verification: bool,
+    /// Initial protocol version
+    #[serde(default = "PlatformTestConfig::default_initial_protocol_version")]
+    pub initial_protocol_version: ProtocolVersion,
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/config.rs]: Reverted back

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs [hunk 10]
```diff
@@ -122,6 +140,87 @@ where
...
+                .expect("next_epoch_activation_date must always be in the range");
+
+            tracing::warn!(
+                next_epoch_protocol_version,
+                latest_supported_protocol_version,
+                "The node doesn't support new protocol version {} that will be activated starting from {}. Please update your software, otherwise the node won't be able to participate in the network.",
+                next_epoch_protocol_version,
+                next_epoch_activation_datetime.to_rfc2822(),
+            );
+        }
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs:222]: In this case, it will be logged only once. I think it's better if we log it every block

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs [hunk 11]
```diff
@@ -45,15 +48,60 @@ where
...
+            if platform_state.current_protocol_version_in_consensus() != next_protocol_version {
+                // TODO: We are calculating epoch info once again which is not great
+                //  but it happens only once on protocol version upgrade (once a month?)
+                epoch_info = self.gather_epoch_info(
+                    &block_proposal,
+                    transaction,
+                    platform_state,
+                    next_platform_version,
+                )?;
+            }
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs:97]: Yeah, you're right. Better I remove it.

File: packages/rs-drive-abci/src/execution/engine/finalize_block_proposal/v0/mod.rs [hunk 12]
```diff
@@ -247,6 +239,8 @@ where
         }
         .into();
 
+        self.update_drive_cache(&block_execution_context, platform_version)?;
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/finalize_block_proposal/v0/mod.rs:242]: We consume BlockExecutionContext to call update_state_cache. If you look at the update_state_cache function you will see that we aren't using drive cache there and not supposed to since update_state_cache should deal with only state cache.

File: packages/rs-drive-abci/tests/strategy_tests/main.rs [hunk 13]
```diff
@@ -1144,7 +1144,7 @@ mod tests {
                     .unwrap()
                     .unwrap()
             ),
-            "8e1e8d1ae51b3fc8a9e4acd0d312c9494943f2b7a5b957cc1379ab246ebd678d".to_string()
+            "aae3eb47619991d0260a5e168bafe28747c6262628bdf7599649bdacf76bc220".to_string()
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/tests/strategy_tests/main.rs:1147]: Because now we save for 1 version propose vote more

File: packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/check_for_desired_protocol_upgrade/v0/mod.rs [hunk 14]
```diff
@@ -1,36 +1,31 @@
...
-        current_protocol_version_in_consensus: ProtocolVersion,
-        transaction: &Transaction,
     ) -> Result<Option<ProtocolVersion>, Error> {
         let required_upgraded_hpns = 1
             + (total_hpmns as u64)
-                .checked_mul(PROTOCOL_VERSION_UPGRADE_PERCENTAGE_NEEDED)
+                .checked_mul(
+                    self.config
+                        .execution
+                        .protocol_version_upgrade_percentage_needed,
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/check_for_desired_protocol_upgrade/v0/mod.rs:19]: Move to PlatfromVersion tree?

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs [hunk 15]
```diff
@@ -106,6 +123,64 @@ where
...
+                last_committed_platform_state,
+                &mut block_platform_state,
+                transaction,
+                platform_version,
+            )?;
+        } else if current_block_protocol_version != previous_block_protocol_version {
+            // It might happen only in case of error in the code.
+            return Err(Error::Execution(
+                ExecutionError::UnexpectedProtocolVersionUpgrade,
+            ));
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs]: Done

File: packages/rs-drive-abci/src/execution/platform_events/epoch/gather_epoch_info/mod.rs [hunk 16]
```diff
@@ -31,14 +31,20 @@ impl<C> Platform<C> {
         transaction: &Transaction,
         platform_state: &PlatformState,
         platform_version: &PlatformVersion,
-    ) -> Result<EpochInfoV0, Error> {
+    ) -> Result<EpochInfo, Error> {
+        // Please be aware epoch information is gathered with previous platform version
+        // on epoch change (1st block of the epoch), despite we are switching to a new version
+        // in this block. A new version of this method will be called for the rest of epoch blocks
+        // and first block of the next epoch.
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/epoch/gather_epoch_info/mod.rs:40]: Done

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs [hunk 17]
```diff
@@ -122,6 +140,87 @@ where
...
         );
 
+        // Update protocol version for this block
+        let previous_block_protocol_version = last_committed_platform_state
+            .current_platform_version()?
+            .protocol_version;
+        let current_block_protocol_version = platform_version.protocol_version;
+
+        block_platform_state
+            .set_current_protocol_version_in_consensus(current_block_protocol_version);
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs]: Done

File: packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/check_for_desired_protocol_upgrade/v0/mod.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/check_for_desired_protocol_upgrade/v0/mod.rs:19]: Okay

File: packages/rs-drive-abci/src/execution/engine/finalize_block_proposal/v0/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/engine/finalize_block_proposal/v0/mod.rs:242]: Okay I reviewed the order change and we *should* be okay because of the height guard for queries.

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs [hunk 18]
```diff
@@ -45,15 +48,50 @@ where
...
     {
-        let platform_version = platform_state.current_platform_version()?;
+        // Epoch information is always calculated with the last committed platform version
+        // even if we are switching to a new version in this block.
+        let last_committed_platform_version = platform_state.current_platform_version()?;
 
-        let epoch_info = self.gather_epoch_info(
+        // !!!! This EpochInfo is based on the last committed platform version
+        // !!!! and will be used for the first block of the epoch.
+        let mut epoch_info = self.gather_epoch_info(
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs:57]: no need for this to be mut

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs [hunk 19]
```diff
@@ -122,6 +140,86 @@ where
...
+                .expect("next_epoch_activation_date must always be in the range");
+
+            tracing::warn!(
+                next_epoch_protocol_version,
+                latest_supported_protocol_version,
+                "The node doesn't support new protocol version {} that will be activated starting from {}. Please update your software, otherwise the node won't be able to participate in the network. https://docs.dash.org/platform-protocol-upgrade",
+                next_epoch_protocol_version,
+                next_epoch_activation_datetime.to_rfc2822(),
+            );
+        }
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs:221]: I really don't want this polluting run block proposal

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs (hunk 19 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs:221]: Yes, I will move to a new function

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs (hunk 18 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs]: Fixed

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs (hunk 19 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs]: Done

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/v0/mod.rs]: Moved to a separate function

File: packages/rs-drive-abci/tests/strategy_tests/upgrade_fork_tests.rs [hunk 20]
```diff
@@ -490,6 +502,166 @@ mod tests {
...
+        // Add a new version to upgrade to new protocol version only with one vote
+        const TEST_PROTOCOL_VERSION_4_WITH_1_HPMN_UPGRADE: u32 =
+            (1 << TEST_PROTOCOL_VERSION_SHIFT_BYTES) + 4;
+
+        let mut test_platform_v4 = PLATFORM_V1.clone();
+        test_platform_v4
+            .drive_abci
+            .methods
+            .protocol_upgrade
+            .protocol_version_upgrade_percentage_needed = 1;
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/tests/strategy_tests/upgrade_fork_tests.rs:516]: Why is this part of this test?

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]
The test you implemented is not broad enough and does not match actual network conditions, it's a complete edge case, while it's not a bad idea to keep it, we need a test that actually shows what happens when a node is removed, but not when the quorum size is 1 node, as this will have us be in the wrong quorums.

File: packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/check_for_desired_protocol_upgrade/v0/mod.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/check_for_desired_protocol_upgrade/v0/mod.rs]: Done

File: packages/rs-drive-abci/tests/strategy_tests/upgrade_fork_tests.rs (hunk 20 above)
Comment (shumkov) [reply packages/rs-drive-abci/tests/strategy_tests/upgrade_fork_tests.rs:516]: We agreed to move this constant to the platform version in the comments above. Since `MasternodeListChangesStrategy` can be set only on `run_chain_for_strategy` and it's falling after few more blocks the solution is to change this percentage to 1 instead of 75 so 1 vote is enough to switch to new version.

**Comment by shumkov** [comment]
@QuantumExplorer actually quorum size is 32 and 50 hpmns. I changed for this test the constant that represents the percentage of masternodes for succeeding voting on new version from 75 to 1. It doesn't change the validator set behavior.

**Review by QuantumExplorer — APPROVED** [review]
